<a href="https://colab.research.google.com/github/IntroComputationalPhysics-UNT/more-on-numpy-Cooper-9/blob/main/More_on_NumPy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
np.__version__


# explore NumPy
dir(np)[:20] # first 20 entries


# get help for a specific function
help(np.sin)


# COMPLETE THIS TASK
# Find documentation for np.linspace using '?'
np.linspace?

Help on ufunc:

sin = <ufunc 'sin'>
    sin(x, /, out=None, *, where=True, casting='same_kind', order='K', dtype=None, subok=True[, signature])

    Trigonometric sine, element-wise.

    Parameters
    ----------
    x : array_like
        Angle, in radians (:math:`2 \pi` rad equals 360 degrees).
    out : ndarray, None, or tuple of ndarray and None, optional
        A location into which the result is stored. If provided, it must have
        a shape that the inputs broadcast to. If not provided or None,
        a freshly-allocated array is returned. A tuple (possible only as a
        keyword argument) must have length equal to the number of outputs.
    where : array_like, optional
        This condition is broadcast over the input. At locations where the
        condition is True, the `out` array will be set to the ufunc result.
        Elsewhere, the `out` array will retain its original value.
        Note that if an uninitialized `out` array is created via the default
        ``

In [11]:
# COMPLETE THIS TASK
# Decompose a 5×5 random matrix into (real) symmetric and antisymmetric parts.
# A symmetric matrix A satisfies A = 0.5*(A + A.T);
# An antisymmetric matrix satisfies A = 0.5*(A - A.T);
# The `.T` attribute finds the transpose of the matrix.

R = np.random.randn(5,5)
R_T = R.T
Symmetric = 0.5*(R+R_T);
Antisymmetric = 0.5*(R-R_T);

Symmetric, Antisymmetric


(array([[-1.24241576,  0.04357826, -0.30454575,  0.83628758,  0.16259077],
        [ 0.04357826,  0.6548555 ,  0.45256335,  0.61620943, -1.33503483],
        [-0.30454575,  0.45256335, -0.76261403,  0.21896236,  0.16431738],
        [ 0.83628758,  0.61620943,  0.21896236,  0.9255768 , -0.12445949],
        [ 0.16259077, -1.33503483,  0.16431738, -0.12445949,  2.23717229]]),
 array([[ 0.        ,  0.38754081, -1.02716116,  0.44339994, -0.22921803],
        [-0.38754081,  0.        , -0.36406599,  0.03930136, -0.81160389],
        [ 1.02716116,  0.36406599,  0.        , -1.02096963,  0.50924835],
        [-0.44339994, -0.03930136,  1.02096963,  0.        ,  0.79422767],
        [ 0.22921803,  0.81160389, -0.50924835, -0.79422767,  0.        ]]))

In [45]:
# COMPLETE THESE TASKS
# 1. look into the documentation for `np.matmul`, `np.dot`, `@`, and `*` and explain their difference(s)
# 2. create an example/examples to explain their differences

A = np.array([[1,2],[3,4]])
B = np.array([[4,5], [6,7]])
C = np.array([[1,5], [6,2]])
D = np.random.randn(1,1)
E = np.random.randn(1,1)


In [38]:
np.matmul(A,B)

array([[16, 19],
       [36, 43]])

In [36]:
A @ C

array([[13,  9],
       [27, 23]])

In [40]:
A * B

array([[ 4, 10],
       [18, 28]])

In [46]:
np.dot(D,E)

array([[0.23877757]])

1. np.matmul is only a matrix product of two arrays (no scalar values).
2. np.dot is also a product of two arrays but can change actions depending on the dimension of the arrays.(more flexible, general dot product)
3. The * operator performs element-wise multiplication (also known as Hadamard product) between two arrays.
4. @ is the 'infix equivalent' to np.matmul (same thing essentially)

In [ ]:
# COMPLETE THIS TASK
# create a vectorized function that calculates the total energy (T+U; kinetic + potential energy) of a harmonic oscillator

In [58]:
def total_harmonic_oscillator_energy(x, v, m, k):
    """
    Calculates the total energy (kinetic + potential) of a harmonic oscillator.

    Parameters:
    x (array_like): Displacement from equilibrium.
    v (array_like): Velocity.
    m (float): Mass of the oscillating object.
    k (float): Spring constant.

    Returns:
    array_like: Total energy of the harmonic oscillator.
    """
    kinetic_energy = 0.5 * m * v**2
    potential_energy = 0.5 * k * x**2
    total_energy = kinetic_energy + potential_energy
    return total_energy

# Define parameters
mass = 1.0  # kg
spring_constant = 10.0  # N/m

# Create arrays for displacement and velocity
displacements = np.array([0.1, 0.2, 0.3, -0.1, -0.2]) # meters
velocities = np.array([0.5, 0.3, 0.1, -0.4, -0.6]) # m/s

# Calculate total energy
energies = total_harmonic_oscillator_energy(displacements, velocities, mass, spring_constant)

print("Displacements (x):", displacements)
print("Velocities (v):", velocities)
print("Total Energies:", energies)

Displacements (x): [ 0.1  0.2  0.3 -0.1 -0.2]
Velocities (v): [ 0.5  0.3  0.1 -0.4 -0.6]
Total Energies: [0.175 0.245 0.455 0.13  0.38 ]


In [61]:
# COMPLETE THIS TASK
# In the above cell, which eigenvector is associated with which eigenvalue?

K = np.array([[ 1,-1, 0]
             ,[-1, 2,-1]
             ,[ 0,-1, 1] ])
vals, vecs = np.linalg.eigh(K)
vals, vecs

(array([9.99658224e-17, 1.00000000e+00, 3.00000000e+00]),
 array([[-5.77350269e-01, -7.07106781e-01,  4.08248290e-01],
        [-5.77350269e-01,  9.71445147e-17, -8.16496581e-01],
        [-5.77350269e-01,  7.07106781e-01,  4.08248290e-01]]))

9.99e-17 is essentially 0, meaning the eigenvalues are 0, 1, and 3. These correspond to the columns of the 'vecs' matrix. (0 to the first, 1 to the second, and 3 to the third)

In [69]:
# COMPLETE THESE TASKS
# Show that U@U.T and U.T@U are 2x2 identity matrices
# Note that normally we would need to use the `.conj().T` attribute to find the conjugate transpose matric, but our matrix is real.

U = np.array([[1,0],[0,1]])
U_T = U.T

print(U @ U.T)
print(U.T @ U)


[[1 0]
 [0 1]]
[[1 0]
 [0 1]]


In [70]:
# COMPLETE THIS TASK
# From the eigenvalue equation M v = λ v, it follows that for U = [v1, v2], then M U = λ U.
# Because λ is diagonal and U.T@U = np.eye(2), it follows that λ = U.T @ K @ U.
# Confirm this behavior.


# U is the matrix whose columns are the eigenvectors
U_eigenvectors = vecs

# Form the diagonal matrix of eigenvalues (lambda_matrix)
lambda_matrix = np.diag(vals)

# Calculate U.T @ K @ U
computed_lambda = U_eigenvectors.T @ K @ U_eigenvectors

print("Original Eigenvalues (diagonal matrix):\n", lambda_matrix)
print("\nComputed U.T @ K @ U:\n", computed_lambda)

# Confirm that the computed_lambda is approximately equal to lambda_matrix
# np.allclose is used for floating-point comparisons
is_confirmed = np.allclose(computed_lambda, lambda_matrix)
print(f"\nDoes U.T @ K @ U approximately equal the diagonal eigenvalue matrix? {is_confirmed}")

# Also confirm that U.T @ U is approximately an identity matrix (for orthogonal eigenvectors)
# This is true for symmetric matrices like K
identity_check = U_eigenvectors.T @ U_eigenvectors
print("\nU.T @ U (Identity check):\n", identity_check)
print(f"Is U.T @ U approximately an identity matrix? {np.allclose(identity_check, np.eye(K.shape[0]))}")

Original Eigenvalues (diagonal matrix):
 [[9.99658224e-17 0.00000000e+00 0.00000000e+00]
 [0.00000000e+00 1.00000000e+00 0.00000000e+00]
 [0.00000000e+00 0.00000000e+00 3.00000000e+00]]

Computed U.T @ K @ U:
 [[ 1.23259516e-32 -7.85046229e-17 -1.35973996e-16]
 [-8.00622087e-17  1.00000000e+00 -6.57248007e-17]
 [-1.98335024e-16 -1.07187350e-16  3.00000000e+00]]

Does U.T @ K @ U approximately equal the diagonal eigenvalue matrix? True

U.T @ U (Identity check):
 [[ 1.00000000e+00 -1.44160965e-16 -2.55774521e-17]
 [-1.44160965e-16  1.00000000e+00  1.46133318e-16]
 [-2.55774521e-17  1.46133318e-16  1.00000000e+00]]
Is U.T @ U approximately an identity matrix? True


In [72]:
# COMPLETE THIS TASK
# Use the eigenvectors to eigenvalues to interpret the two modes in this model of a diatomic molecule

In [73]:
k = 2450
K = np.array([[k, -k], [-k, k]])

vals, vecs = np.linalg.eigh(K)

omega = np.sqrt(vals)

print("Force constant (k):", k)
print("K matrix:\n", K)
print("Eigenvalues (vals):", vals)
print("Eigenvectors (vecs):\n", vecs)
print("Vibrational frequencies (omega):", omega)

Force constant (k): 2450
K matrix:
 [[ 2450 -2450]
 [-2450  2450]]
Eigenvalues (vals): [   0. 4900.]
Eigenvectors (vecs):
 [[-0.70710678 -0.70710678]
 [-0.70710678  0.70710678]]
Vibrational frequencies (omega): [ 0. 70.]


### Explanation of Potential Energy Formula

The potential energy formula for a system with displacements `x` and force constant matrix `K` is given by `V = 0.5 * x.T @ K @ x`. In the context of a diatomic molecule, `x = (x1, x2)` represents the displacements of the two atoms from their equilibrium positions. `x.T` is the transpose of the displacement vector.

This formula calculates the total potential energy stored in the system due to the elastic deformation (stretching or compression) of the bond between the two atoms. The `K` matrix captures the interatomic forces. For a diatomic molecule, the `K` matrix describes how the potential energy changes with the relative displacement of the two atoms. A higher potential energy indicates a greater 'strain' or displacement from the equilibrium configuration.

### Interpretation of the First Eigenvalue and Eigenvector

**Eigenvalue 1 (λ₁):** `vals[0] = 0` (or `9.99658224e-17` from the previous general problem, but for this specific diatomic molecule it's exactly 0).

**Corresponding Eigenvector 1 (v₁):** `vecs[:, 0] = [-0.70710678, -0.70710678]`

**Vibrational Frequency 1 (ω₁):** `omega[0] = np.sqrt(0) = 0`

**Physical Interpretation:**

1.  **Motion of Atoms:** The eigenvector `[-0.70710678, -0.70710678]` indicates that both atoms are displaced in the same direction and by the same magnitude. For instance, if the first atom moves left by 0.707 units, the second atom also moves left by 0.707 units (or vice versa, depending on the chosen direction for the eigenvector, as `[0.707, 0.707]` would also be a valid eigenvector). This represents a **pure translational motion** of the entire molecule as a single unit, where the bond length between the two atoms remains unchanged.

2.  **Vibrational Frequency and Potential Energy:** The corresponding vibrational frequency `ω₁ = 0` is crucial. A zero frequency implies that this mode requires **no energy** to be excited and does not involve any oscillation. If the molecule is translated, there is no change in the potential energy. This is consistent with the potential energy formula `V = 0.5 * x.T @ K @ x`. If `x` is proportional to `v₁`, then `K v₁ = λ₁ v₁ = 0 * v₁ = 0`, so `V = 0.5 * x.T @ 0 = 0`. This means that translating the molecule as a whole does not stretch or compress the bond, and thus stores no potential energy in the internal spring.

3.  **Mode Type:** This mode is a **translational mode**, not a vibrational mode. It describes the motion of the center of mass of the diatomic molecule without any internal deformation.

### Interpretation of the Second Eigenvalue and Eigenvector

**Eigenvalue 2 (λ₂):** `vals[1] = 4900`

**Corresponding Eigenvector 2 (v₂):** `vecs[:, 1] = [-0.70710678, 0.70710678]`

**Vibrational Frequency 2 (ω₂):** `omega[1] = np.sqrt(4900) = 70`

**Physical Interpretation:**

1.  **Motion of Atoms:** The eigenvector `[-0.70710678, 0.70710678]` indicates that the two atoms are displaced in **opposite directions** but with the **same magnitude**. If the first atom moves left by 0.707 units, the second atom moves right by 0.707 units. This represents a **relative motion** where the distance between the two atoms changes (the bond is stretched or compressed).

2.  **Vibrational Frequency and Potential Energy:** The corresponding vibrational frequency `ω₂ = 70` is non-zero. A non-zero frequency signifies that this mode is a true oscillation and requires energy to be excited. When the atoms move according to this eigenvector, the bond length changes, leading to a change in the potential energy of the system. This is consistent with `V = 0.5 * x.T @ K @ x`. If `x` is proportional to `v₂`, then `K v₂ = λ₂ v₂`. Since `λ₂` is non-zero, the potential energy `V` will also be non-zero, indicating energy stored in the deformation.

3.  **Mode Type:** This mode is a **vibrational mode**. It describes the internal oscillation of the diatomic molecule, where the atoms move relative to each other, stretching and compressing the bond.

### Summary of Normal Modes for the Diatomic Molecule

For the diatomic molecule described by the `K` matrix `[[k, -k], [-k, k]]` (with `k = 2450`), we have identified two normal modes:

1.  **First Normal Mode (Translational Mode):**
    *   **Eigenvalue (λ₁):** `0`
    *   **Eigenvector (v₁):** `[-0.70710678, -0.70710678]` (or `[1, 1]` normalized)
    *   **Vibrational Frequency (ω₁):** `0`
    *   **Physical Meaning:** This mode represents the **center-of-mass translation** of the entire diatomic molecule. Both atoms move together in the same direction and with the same magnitude. Since there is no change in the relative distance between the atoms, the bond length remains constant, and therefore, no potential energy is stored in this motion (`V = 0`). The zero vibrational frequency signifies that this mode is not an oscillation but a free translation, which does not require energy to be excited.

2.  **Second Normal Mode (Vibrational Mode):**
    *   **Eigenvalue (λ₂):** `4900`
    *   **Eigenvector (v₂):** `[-0.70710678, 0.70710678]` (or `[-1, 1]` normalized)
    *   **Vibrational Frequency (ω₂):** `70`
    *   **Physical Meaning:** This mode represents the **internal vibration** of the diatomic molecule. The two atoms move in opposite directions, stretching and compressing the bond. This relative motion changes the distance between the atoms, leading to a storage of potential energy (`V > 0`). The non-zero vibrational frequency (70 rad/s) indicates that this is a true oscillation, and energy is required to excite this mode. This is the only true vibrational mode for a linear diatomic molecule.

#**I didn't exactly know what the last question was asking, so hopefully gemini was able to give a sufficient answer.**